.travis.yml

```diff
@@ -14,24 +14,147 @@ git:
+    -
+      <<: *foursix
+      env: GCC=-4.6 MPI=MPICH
+    - &fourseven
+      <<: *baseconf
+      addons:
+        apt:
+          sources: ['ubuntu-toolchain-r-test']
+          packages: ['gcc-4.7', 'g++-4.7', 'gfortran-4.7', 'ccache']
+      env: GCC=-4.7 MPI=none
+    -
+      <<: *fourseven
+      env: GCC=-4.7 MPI=MPICH
+    - &foureight
+      <<: *baseconf
+      addons:
+        apt:
+          sources: ['ubuntu-toolchain-r-test']
+          packages: ['gcc-4.8', 'g++-4.8', 'gfortran-4.8', 'ccache']
+      env: GCC=-4.8 MPI=none
+    -
+      <<: *foureight
+      env: GCC=-4.8 MPI=MPICH
+    - &fournine
+      <<: *baseconf
+      addons:
+        apt:
+          sources: ['ubuntu-toolchain-r-test']
+          packages: ['gcc-4.9', 'g++-4.9', 'gfortran-4.9', 'ccache']
+      env: GCC=-4.9 MPI=none
+    -
+      <<: *fournine
+      env: GCC=-4.9 MPI=MPICH
+    - &five
+      <<: *baseconf
+      addons:
+        apt:
+          sources: ['ubuntu-toolchain-r-test']
+          packages: ['gcc-5', 'g++-5', 'gfortran-5', 'ccache']
+      env: GCC=-5 MPI=none
+
+    -
+      <<: *five
+      env: GCC=-5 MPI=MPICH
+    - &six
+      <<: *baseconf
+      addons:
+        apt:
+          sources: ['ubuntu-toolchain-r-test']
+          packages: ['gcc-6', 'g++-6', 'gfortran-6', 'ccache']
+      env: GCC=-6 MPI=none
+    -
+      <<: *six
+      env: GCC=-6 MPI=MPICH
 
 before_install:
-  - openssl aes-256-cbc -K $encrypted_c48b4c75cf28_key -iv $encrypted_c48b4c75cf28_iv -in id_rsa_Travis2GH.enc -out id_rsa_Travis2GH -d # encrypted with `travis encrypt-file`
-  - unset -v ${!encrypted_@} # unset secret encrypted variables
-  - chmod 600 id_rsa_Travis2GH
+  - if [[ ! -d "${HOME}/.ccache" ]] ; then mkdir "${HOME}/.ccache" ; else ls -a "${HOME}/.ccache" ; fi
+  - ccache --zero-stats
   - if [[ ! -d "${HOME}/bin" ]] ; then mkdir "${HOME}/bin" && export PATH="${HOME}/bin:${PATH}"; fi
-  - printf '#!/usr/bin/env bash\nssh -i %s/id_rsa_Travis2GH "$@"\n' "$(pwd)" > "${HOME}/bin/ssh_to_GH"
-  - chmod +x "${HOME}/bin/ssh_to_GH"
-  - export GIT_SSH="${HOME}/bin/ssh_to_GH"
-  - echo "Testing ssh to github.com" && ($GIT_SSH -T "git@github.com" || true)
+  - if [[ ! -d "${HOME}/mpich/build" ]] ; then mkdir -p "${HOME}/mpich/build" ; fi
+  - if [[ ! -d "${HOME}/usr" ]] ; then mkdir -p "${HOME}/usr/{bin,lib,share,include,lib64,libexec}" && export PATH="${HOME}/usr/bin:${PATH}"; fi
+  - |
+    if [[ -n "${encrypted_c48b4c75cf28_key}" ]] ; then # not a PR or someone else's repo
+      openssl aes-256-cbc -K $encrypted_c48b4c75cf28_key -iv $encrypted_c48b4c75cf28_iv -in id_rsa_Travis2GH.enc -out id_rsa_Travis2GH -d \
+      && chmod 600 id_rsa_Travis2GH \
+      && printf '#!/usr/bin/env bash\nssh -i %s/id_rsa_Travis2GH "$@"\n' "$(pwd)" > "${HOME}/bin/ssh_to_GH" \
+      && chmod +x "${HOME}/bin/ssh_to_GH" \
+      && export GIT_SSH="${HOME}/bin/ssh_to_GH" \
+      && echo "Testing ssh to github.com" && ($GIT_SSH -T "git@github.com" || true)
+    fi
```

Comment:
This should be ported to master: don't try to decrypt when in a pull request or on someone else's fork (I worked from my fork to reduce Travis-CI noise, after realizing I would be pushing a lot of broken commits)